json-endpoints/jetpack/class.jetpack-json-api-plugins-install-endpoint.php

```diff
@@ -81,3 +90,111 @@ protected static function get_plugin_id_by_slug( $slug ) {
 		return false;
 	}
 }
+/**
+ * Allows us to capture that the site doesn't have proper file system access.
+ * In order to update the plugin.
+ */
+class Jetpack_Automatic_Plugin_Install_Skin extends Automatic_Upgrader_Skin {
+	/**
+	 * Stores the last error key;
+	 **/
+	protected $main_error_code = null;
+
+	/**
+	 * Stores the last error message.
+	 **/
+	protected $main_error_message = null;
+
+	/**
+	 * Overwrites the set_upgrader to be able to tell if we e ven have the ability to write to the files.
+	 *
+	 * @param WP_Upgrader $upgrader
+	 *
+	 */
+	public function set_upgrader( &$upgrader ) {
+		parent::set_upgrader( $upgrader );
+
+		// Check if we even have permission to.
+		$result = $upgrader->fs_connect( array( WP_CONTENT_DIR, WP_PLUGIN_DIR ) );
+		if( !$result ) {
+			// set the string here since they are not available just yet
+			$upgrader->generic_strings();
+			$this->feedback( 'fs_unavailable' );
+		}
+	}
+
+	/**
+	 * Overwrites the error function
+	 */
+	public function error( $error ) {
+		if(	is_wp_error( $error ) ) {
+			$this->feedback( $error );
+		}
+	}
+
+	private function set_main_error_code( $code ) {
+		// Don't set the process_failed as code since it is not that helpful unless we don't have one already set.
+		$this->main_error_code = ( $code == 'process_failed' && $this->main_error_code  ? $this->main_error_code : $code );
+	}
+
+	private function set_main_error_message( $message, $code ) {
+		// Don't set the process_failed as message since it is not that helpful unless we don't have one already set.
+		$this->main_error_message = ( $code == 'process_failed' && $this->main_error_code ? $this->main_error_code : $message );
+	}
+
+	public function get_main_error_code() {
+		return $this->main_error_code ? $this->main_error_code : 'install_error';
```

Comment:
Instead of doing the ternary operator we could make the attributes public and assign `install_error` by default: `public $main_error_code = 'install_error';` what do you think?